File: neetcode150-py/src/neetcode150_py/arrays_and_hashing/q271_encode_and_decode_strings.py
from typing import List


class Solution:
    def encode(self, strs: List[str]) -> str:
        encoded = ""
        for s in strs:
            encoded += str(len(s)) + "#" + s
        return encoded

    def decode(self, s: str) -> List[str]:
        decoded: List[str] = []
        i = 0

        while i < len(s):
            j = i
            while s[j] != "#":
                j += 1
            wordlen = int(s[i:j])
            decoded.append(s[j + 1 : j + 1 + wordlen])
            i = j + wordlen + 1
        return decoded

    # def decode_(self, s: str) -> List[str]:
    #     decoded: List[str] = []

    #     if len(s) == 0:
    #         return decoded

    #     i = 0
    #     while i < len(s):
    #         wordlen_str = ""
    #         while i < len(s) and s[i].isdigit():
    #             wordlen_str += s[i]
    #             i += 1
    #         if s[i] == "#":
    #             i += 1

    #         wordlen = int(wordlen_str)
    #         decoded.append(s[i : i + wordlen])
    #         i += wordlen
    #     return decoded

File: neetcode150-rs/src/arrays_and_hashing/q271_encode_and_decode_strings.rs
pub struct Solution;

impl Solution {
    pub fn encode(strs: Vec<String>) -> String {
        strs.into_iter()
            .map(|s| format!("{}#{}", s.len(), s))
            .collect()
    }

    pub fn decode(s: String) -> Vec<String> {
        // Look for digit#
        let mut decoded: Vec<String> = Vec::new();
        let mut i = 0;
        let chars: Vec<char> = s.chars().collect();

        while i < chars.len() {
            let mut j = i;
            while chars[j] != '#' {
                j += 1;
            }

            let wordlen: usize = chars[i..j].iter().collect::<String>().parse().unwrap();
            j += 1; // skip the '#'

            decoded.push(chars[j..j + wordlen].iter().collect());
            i = j + wordlen;
        }
        decoded
    }
}

#[cfg(test)]
mod tests {
    use crate::arrays_and_hashing::q271_encode_and_decode_strings::Solution;

    #[test]
    fn test_encode_and_decode() {
        let strs = vec!["hello".to_string(), "world".to_string()];
        let encoded = Solution::encode(strs.clone());
        let decoded = Solution::decode(encoded);
        assert_eq!(decoded, strs);
    }

    #[test]
    fn test_empty_strings() {
        let strs = vec!["".to_string(), "test".to_string()];
        let encoded = Solution::encode(strs.clone());
        let decoded = Solution::decode(encoded);
        assert_eq!(decoded, strs);
    }

    #[test]
    fn test_example1() {
        let strs: Vec<String> = vec!["neet", "code", "love", "you"]
            .iter()
            .map(|s| s.to_string())
            .collect();
        let encoded = Solution::encode(strs.clone());
        let decoded = Solution::decode(encoded);
        assert_eq!(decoded, strs);
    }

    #[test]
    fn test_example2() {
        let strs: Vec<String> = vec!["we", "say", ":", "yes"]
            .iter()
            .map(|s| s.to_string())
            .collect();
        let encoded = Solution::encode(strs.clone());
        let decoded = Solution::decode(encoded);
        assert_eq!(decoded, strs);
    }
}
